### PR TITLE
Fix init of WAL page header at startup

### PR DESCRIPTION
**Pull request by hlinnaka**
If the primary is started at an LSN within the first of a 16 MB WAL segment, the "long XLOG page header" at the beginning of the segment was not initialized correctly. That has gone unnnoticed, because under normal circumstances, nothing looks at the page header. The WAL that is streamed to the safekeepers starts at the new record's LSN, not at the beginning of the page, so that bogus page header didn't propagate elsewhere, and a primary server doesn't normally read the WAL its written. Which is good because the contents of the page would be bogus anyway, as it wouldn't contain any of the records before the LSN where the new record is written.

Except that in the following cases a primary does read its own WAL:

1. When there are two-phase transactions in prepared state at checkpoint. The checkpointer reads the two-phase state from the XLOG_XACT_PREPARE record, and writes it to a file in pg_twophase/.

2. Logical decoding reads the WAL starting from the replication slot's restart LSN.

This PR fixes the problem with two-phase transactions. For that, it's sufficient to initialize the page header correctly. The checkpointer only needs to read XLOG_XACT_PREPARE records that were generated after the server startup, so it's still OK that older WAL is missing / bogus.

I have not investigated if we have a problem with logical decoding, however. Let's deal with that separately.

### CODE REVIEW

**Comment by hlinnaka** [comment]
I happened to notice this bug while I was writing a new test case for an issue with prepared transactions on the v17 branch. It turned out to be a pre-existing bug.

**Comment by hlinnaka** [comment]
> @hlinnaka is the target branch the one you want this to be merged into? It seems like a fairly arbitrary branch to me.

Right, it was just because I made those other change first and knew they were going to be merged soon. It's been merged now, so the point is moot.

**Comment by clarkbw** [comment]
Leaving in reminder for this here: https://github.com/neondatabase/neon/pull/8916/files#diff-116a0da838509394db2ba0a7c6bec4c72808bb993040feba0bfde1e17fc4a93eR23

File: vendor/revisions.json [hunk 1]
```diff
@@ -1,14 +1,14 @@
 {
   "v16": [
     "16.4",
-    "6e9a4ff6249ac02b8175054b7b3f7dfb198be48b"
+    "e6203b5b66f8639461301a980e7a2fb1f9f7562f"
```
Comment (ololobus) [inline vendor/revisions.json:4]: @hlinnaka what's the status of this PR? I see that only 14-16 versions were bumped. Should we do the same for PG 17?

Comment (hlinnaka) [reply vendor/revisions.json]: yep, added that

**Comment by hlinnaka** [comment]
Corresponding PRs in the postgres repo:

v17: https://github.com/neondatabase/postgres/pull/506
v16: https://github.com/neondatabase/postgres/pull/481
v15: https://github.com/neondatabase/postgres/pull/482
v14: https://github.com/neondatabase/postgres/pull/483

**Comment by hlinnaka** [comment]
Rebased, this is now ready to be merged again. I need approval for the v17 PR: https://github.com/neondatabase/postgres/pull/506

**Comment by hlinnaka** [comment]
CI failure on one branch, seems to be flaky https://neon-github-public-dev.s3.amazonaws.com/reports/pr-8914/10957027921/index.html#suites/572721f46de364f1302c76c19e8023c4/c4814c8a998655dd/:

```
PG:2024-09-20 10:41:59.644 GMT [313604] LOG:  [WP] got VoteResponse from acceptor 127.0.0.1:18364, voteGiven=1, epoch=2, flushLsn=0/2000000, truncateLsn=0/14EE280, timelineStartLsn=0/14EE280
PG:2024-09-20 10:41:59.644 GMT [313604] LOG:  [WP] got votes from majority (1) of nodes, term 3, epochStartLsn 0/2000000, donor 127.0.0.1:18364, truncate_lsn 0/14EE280
2024-09-20T10:41:59.668726Z  INFO starting activity monitor for postgresql://cloud_admin@127.0.0.1:18368/postgres?application_name=compute_activity_monitor
PG:2024-09-20 10:41:59.644 GMT [313604] PANIC:  [WP] collected propEpochStartLsn 0/2000000, but basebackup LSN 0/14EE280
PG:2024-09-20 10:41:59.673 GMT [313607] LOG:  [NEON_SMGR] [shard 0] libpagestore: connected to 'postgresql://no_user@localhost:18361' with protocol version 2
PG:2024-09-20 10:41:59.753 GMT [313594] LOG:  background worker "WAL proposer" (PID 313604) was terminated by signal 6: Aborted
```

@arssher is this a known issue?

**Comment by hlinnaka** [comment]
> CI failure on one branch, seems to be flaky https://neon-github-public-dev.s3.amazonaws.com/reports/pr-8914/10957027921/index.html#suites/572721f46de364f1302c76c19e8023c4/c4814c8a998655dd/:
> 
> ```
> PG:2024-09-20 10:41:59.644 GMT [313604] LOG:  [WP] got VoteResponse from acceptor 127.0.0.1:18364, voteGiven=1, epoch=2, flushLsn=0/2000000, truncateLsn=0/14EE280, timelineStartLsn=0/14EE280
> PG:2024-09-20 10:41:59.644 GMT [313604] LOG:  [WP] got votes from majority (1) of nodes, term 3, epochStartLsn 0/2000000, donor 127.0.0.1:18364, truncate_lsn 0/14EE280
> 2024-09-20T10:41:59.668726Z  INFO starting activity monitor for postgresql://cloud_admin@127.0.0.1:18368/postgres?application_name=compute_activity_monitor
> PG:2024-09-20 10:41:59.644 GMT [313604] PANIC:  [WP] collected propEpochStartLsn 0/2000000, but basebackup LSN 0/14EE280
> PG:2024-09-20 10:41:59.673 GMT [313607] LOG:  [NEON_SMGR] [shard 0] libpagestore: connected to 'postgresql://no_user@localhost:18361' with protocol version 2
> PG:2024-09-20 10:41:59.753 GMT [313594] LOG:  background worker "WAL proposer" (PID 313604) was terminated by signal 6: Aborted
> ```
> 
> @arssher is this a known issue?

The case is basically just:

1. Create branch, start an endpoint on it
2. SELECT pg_switch_wal()
3. Immediately stop the endpoint
4. Start a new endpoint

**Comment by arssher** [comment]
> is this a known issue?

I haven't seen exactly this sequence with xlog switch. But it is kind of expected behaviour. CI failure misses first ep logs (it is better to do `endpoint.stop(mode="immediate")` to preserve them instead of `endpoint.stop_and_destroy()`), but I think the following happens:
1) SELECT pg_switch_wal() is done at first compute;
2) first compute is SIGKILL'ed
3) second compute starts and observers that sk position is 0/14EE280, nothing to sync, and fetches basebackup at this pos
4) safekeeper gets from TCP stream WAL switch record and bumps flush position to 0/2000000
5) basebackup is now spoiled because it is taken at wrong LSN position.

So I'd suggest to add sleep/retry until endpoint starts well. Alternatively safekeeper exposes list of walreceivers and we can poll sk until walreceiver on it is gone.

**Comment by hlinnaka** [comment]
> > is this a known issue?
> 
> I haven't seen exactly this sequence with xlog switch, but it is kind of expected behaviour. 

I was able to reproduce this locally with:
```
    for i in range(1, 2000):
        branch_name = f"test_twophase{i}"
        env.neon_cli.create_branch(branch_name)
        endpoint = env.endpoints.create_start(branch_name)
        endpoint.safe_psql("SELECT pg_switch_wal()")
        endpoint.stop_and_destroy()
        endpoint = env.endpoints.create_start(branch_name)
        endpoint.safe_psql("SELECT pg_switch_wal()")
        endpoint.stop_and_destroy()
```

failed after about ~100 iterations

> CI failure misses first ep logs (it is better to do `endpoint.stop(mode="immediate")` to preserve them instead of `endpoint.stop_and_destroy()`), but I think the following happens:
> 
>     1. SELECT pg_switch_wal() is done at first compute;
> 
>     2. first compute is SIGKILL'ed
> 
>     3. second compute starts and observers that sk position is 0/14EE280, nothing to sync, and fetches basebackup at this pos
> 
>     4. safekeeper gets from TCP stream WAL switch record and bumps flush position to 0/2000000
> 
>     5. basebackup is now spoiled because it is taken at wrong LSN position.
> 
> 
> So I'd suggest to add sleep/retry until endpoint starts well. Alternatively safekeeper exposes list of walreceivers and we can poll sk until walreceiver on it is gone.

Hmm, isn't this a potential problem in production too?

**Comment by hlinnaka** [comment]
Opened https://github.com/neondatabase/neon/issues/9079 to track that flakiness, since it seems unrelated to this PR. In this PR, to work around this, I'll add a call to wait for the WAL to be flushed before stopping the compute.

**Pull request merged**